engine_modules/market/tasks.py

```diff
@@ -1,8 +1,127 @@
 # -*- coding: utf-8 -*-
+from django.db.models import Max
+
 from engine.tasks import ResolutionTask
 from engine_modules.market.models import CorporationMarket
 
 
+class UpdateBubblesTask(ResolutionTask):
+	"""
+	Update the bubble value on the CorporationMarket objects
+	"""
+	RESOLUTION_ORDER = 500
+
+	def run(self, game):
+		# We can force the use of bubbles  using the force_bubbles flag
+
+		# I do NOT remember why the test was "not hasattr(game, 'force_bubbles')", I left it here just in case
+		# but the check == False is necessary. TODO: check if 'hasattr' has meaning (-> can a game not have an attribute 'force_bubbles' ?)
+		if game.disable_side_effects and (not hasattr(game, 'force_bubbles') or not game.force_bubbles):
+			return
+
+		corporation_markets = CorporationMarket.objects.filter(corporation__game=game, turn=game.current_turn).annotate(maxval=Max('value')).order_by('market__name', '-value')
+		modifiers = {}
+		markets_blacklist = []
+		for i in range(len(corporation_markets)):
+			cm = corporation_markets[i]
+			if cm.corporation not in modifiers.keys():
+				modifiers[cm.corporation] = 0
+			if cm.market not in markets_blacklist:
+				if cm.value == 0:
+					# Negative bubble
+					cm.bubble_value = -1
+					cm.save()
+					# We have to propagate the modification to the corporation's assets_modifier
+					modifiers[cm.corporation] += cm.bubble_value
+				# This 'elif' may appear innocuous, but by no means is it so: it signifies that a CorporationMarket with a value at 0 is NOT eligible
+				# to a dominaion bubble, even if other Corporations were to have even lower assets on this Market.
+				# This might seem redundant because Markets are not supposed to have negative values, but if taht changes (unlikely), this'll be ready.
+				# In any case, tha means '0' CANNOT be a superposed state for 'bubble_value', so we don't need 2 fields (for positive and negative bubbles)
+				elif cm.value == cm.maxval:
+					if (i == len(corporation_markets) - 1) or (corporation_markets[i + 1].market != cm.market) or (corporation_markets[i + 1].value < cm.value):
+						cm.bubble_value = 1
+						cm.save()
+						# We have to propagate the modification to the corporation's assets_modifier
+						modifiers[cm.corporation] += cm.bubble_value
+					else:
+						# There can be no domination bubble on this market
+						markets_blacklist.append(cm.market)
```

Comment:
actually, i think all this algorithm could be easier to understand if we were to run 2 SQL queries: one for all the bubbles at 0, and another one to retrieve bubbles at maximum